# Cleanup on Row 7

1. Clean the File
2. Handle Corrupt Rows
3. Explore Data Set Bias

## Import Libraries

In [1]:
import numpy as np
import pandas as pd
import matplotlib
import matplotlib.pyplot as plt

### Helper Display Functions

In [5]:
def display_df(df, nrows=10, ncols=None):
    with pd.option_context('display.max_rows', nrows, 'display.max_columns', ncols):
        display (df)
def print_row(df, row):
    for ctr,i in enumerate(df.iloc[row]):
        print (str(df.columns[ctr])+": "+str(i))

### Helper Statistical Functions

In [18]:
#gets ratio of NaNs for each column
def stats_NaN(df):
    df_stats = pd.DataFrame(index=[df.columns], columns=["NaN Ratio"])
    for col in df.columns:
        df_stats["NaN Ratio"][col] = df[col].isna().sum()/len(df) #NaN ratio
    return df_stats.sort_values(by=['NaN Ratio'])

#helps decide what to make dummies for and what to drop
def stats_unique(df, labels):
    df_counts = pd.DataFrame(index=labels,columns=['Unique Count'])
    for l in labels:
        df_counts['Unique Count'][l] = df[l].nunique()
    return df_counts

## Importing the Data

In [52]:
df_raw = pd.read_csv("../dirty_sample_small.csv")

ParserError: Error tokenizing data. C error: Expected 48 fields in line 652242, saw 49


Some rows have an extra field. We shall investigate what this is. First we will import the file in a more managable way, capable of handling these erroneous lines.

In [55]:
df_raw = pd.DataFrame([line.strip().split(',') for line in open('../dirty_sample_small.csv', 
                                                                'r', encoding="ISO-8859-1")])

In [63]:
len(df_raw)

661487

## Investigating Corrupt Lines

In [50]:
stats_NaN(df_raw)

NaN Ratio
0             0
1             0
2             0
3   0.000270602
4   0.000498876
5   0.000757384
6    0.00102496
7    0.00126231
8    0.00155407
9    0.00181712
10   0.00207563
11   0.00235379
12   0.00264404
13   0.00291767
14   0.00316257
15   0.00343166
16   0.00372192
17   0.00397589
18   0.00421928
19   0.00446116
20   0.00468339
21   0.00495399
22   0.00520948
23   0.00549066
24   0.00576882
25   0.00600616
26   0.00624804
27   0.00652016
28   0.00674238
29   0.00698426
30   0.00722463
31   0.00749977
32   0.00774618
33   0.00801225
34   0.00826471
35   0.00856706
36   0.00882406
37   0.00909617
38   0.00936526
39   0.00962075
40   0.00989891
41   0.00989891
42   0.00989891
43   0.00989891
44   0.00989891
45   0.00989891
46   0.00989891
47   0.00989891
48     0.986152

We see that that the extra has 98.6% field NaNs. This is enough to drop it. However we want to do our due diligence and investigate what the values that are not NaN are. Our suspicion is that it contains the empty string "". So we will query this dataset on the extra field for values that are not NaN or not the empty string. These rows with the empty string are likely corrupt.

In [56]:
df_raw[(pd.notna(df_raw[48]) & (df_raw[48] != ''))][48]

Series([], Name: 48, dtype: object)

We see that this returns nothing. Therefore everything in the extra field is either NaN or the empty string, so there is absolutely no value in keeping it and we will drop it. But first, lets see how many of these corrupt erroneous lines there are. We know that the construction of df_raw line by line as we did above is indifferent to the amount of columns in each row. Most of the time it saw 48 columns, in the later rows it saw 49 because of the extra empty string. The construction of the data set put NaNs in the earlier rows to ensure the shape of the dataset is intact (a regular rectangle). Therefore the source of the corrupt lines is the empty string, so if we count the empty strings in the extra field we will have successfully counted the amount of corrupt lines.

In [65]:
len(df_raw[((df_raw[48] == ''))][48])

9160

We see that there are 9160 of these rows.

In [67]:
display_df(df_raw[((df_raw[48] == ''))])

0        1     2      3     4      5      6   \
652241  HarvardX/CS50x/2014_T1  1818882  True   True  True  False  False   
652242  HarvardX/CS50x/2014_T1  1818882  True   True  True  False  False   
652243  HarvardX/CS50x/2014_T1  1818882  True   True  True  False  False   
652244  HarvardX/CS50x/2014_T1  1818882  True   True  True  False  False   
652245  HarvardX/CS50x/2014_T1  1818882  True   True  True  False  False   
...                        ...      ...   ...    ...   ...    ...    ...   
661482  HarvardX/CS50x/2014_T1  1934738  True  False        False  False   
661483  HarvardX/CS50x/2014_T1  1934738  True  False        False  False   
661484  HarvardX/CS50x/2014_T1  1934738  True  False        False  False   
661485  HarvardX/CS50x/2014_T1  1934738  True  False        False  False   
661486  HarvardX/CS50x/2014_T1  1934738  True  False        False  False   

                    7   8              9         10      11  12  \
652241  173.76.227.129  US  United States  Americas  Newton  MA   
652242  173.76.227.129  US  United States  Americas  Newton  MA   
652243  173.76.227.129  US  United States  Americas  Newton  MA   
652244  173.76.227.129  US  United States  Americas  Newton  MA   
652245  173.76.227.129  US  United States  Americas  Newton  MA   
...                ...  ..            ...       ...     ...  ..   
661482   71.41.229.179  US  United States  Americas  Dallas  TX   
661483   71.41.229.179  US  United States  Americas  Dallas  TX   
661484   71.41.229.179  US  United States  Americas  Dallas  TX   
661485   71.41.229.179  US  United States  Americas  Dallas  TX   
661486   71.41.229.179  US  United States  Americas  Dallas  TX   

                   13     14                15                 16 17 18  \
652241  Massachusetts  02458  Northern America  Developed regions         
652242  Massachusetts  02458  Northern America  Developed regions         
652243  Massachusetts  02458  Northern America  Developed regions         
652244  Massachusetts  02458  Northern America  Developed regions         
652245  Massachusetts  02458  Northern America  Developed regions         
...               ...    ...               ...                ... .. ..   
661482          Texas  75201  Northern America  Developed regions         
661483          Texas  75201  Northern America  Developed regions         
661484          Texas  75201  Northern America  Developed regions         
661485          Texas  75201  Northern America  Developed regions         
661486          Texas  75201  Northern America  Developed regions         

             19        20  21    22 23   24   25                   26  \
652241  42.3533  -71.1883  el  2002  m  0.0  0.6  2013-12-04 04:02:20   
652242  42.3533  -71.1883  el  2002  m  0.0  0.6  2013-12-04 04:02:20   
652243  42.3533  -71.1883  el  2002  m  0.0  0.6  2013-12-04 04:02:20   
652244  42.3533  -71.1883  el  2002  m  0.0  0.6  2013-12-04 04:02:20   
652245  42.3533  -71.1883  el  2002  m  0.0  0.6  2013-12-04 04:02:20   
...         ...       ...  ..   ... ..  ...  ...                  ...   
661482  32.7831  -96.8067   p  1972  m  0.0  0.6  2013-07-09 01:46:41   
661483  32.7831  -96.8067   p  1972  m  0.0  0.6  2013-07-09 01:46:41   
661484  32.7831  -96.8067   p  1972  m  0.0  0.6  2013-07-09 01:46:41   
661485  32.7831  -96.8067   p  1972  m  0.0  0.6  2013-07-09 01:46:41   
661486  32.7831  -96.8067   p  1972  m  0.0  0.6  2013-07-09 01:46:41   

                                27                          28    29  30   31  \
652241  2013-12-04 04:02:20.199004  2014-04-01 23:30:47.988143  2194  39  680   
652242  2013-12-04 04:02:20.199004  2014-04-01 23:30:47.988143  2194  39  680   
652243  2013-12-04 04:02:20.199004  2014-04-01 23:30:47.988143  2194  39  680   
652244  2013-12-04 04:02:20.199004  2014-04-01 23:30:47.988143  2194  39  680   
652245  2013-12-04 04:02:20.199004  2014-04-01 23:30:47.988143  2194  39  680   
...                            ...            

Many of these "corrupt" lines contains valid data, so we will have to hangle. It does appear though that many of these 9160 corrupt lines are indeed duplicates. Lets see how many.

In [72]:
len(df_raw[((df_raw[48] == ''))].drop_duplicates(inplace=False))

710

So of the 9160 corrupt lines, at least 9160 - 710 = 8450 are duplicates. I say at least because even those 710 lines may be duplicates with the rest of the raw dataset as this analysis above only considers rows with the empty string in the extra field. So we can analyze overall the amount of duplicates in the whole raw dataset and compare.

In [73]:
len(df_raw.drop_duplicates(inplace=False))

57223

Interestingly, there are still many more duplicates. However, if we had dropped duplicates at this stage, then we would have also gotten rid of at least 8450 corrupt lines. Therefore we posit, that dropping duplicates would drastically change the count of corrupt lines. We shall prove this with the code below.

In [74]:
len(df_raw.drop_duplicates(inplace=False)[df_raw[48] == ''])

/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  """Entry point for launching an IPython kernel.


710

We see that that these corrupt lines do have 710 valid samples. Therefore simply dropping duplicates in the whole dataset is guarunteed to drop 8450 of the corrupt lines.

## Cleaning the Mixed Up Data

We observe that some of the data appears scrambled. Meaning some of the data for some columns is in the wrong columns. We will have to repair this to have a meaningful dataset. But first, lets give our dataset acceptable column names and drop the erroneous column. From now on, we will be using the goal dataset (the future fully cleaned one) called df_clean.

In [106]:
df_clean = df_raw.copy()
df_clean = df_clean.rename(columns=df_raw.iloc[0])
df_clean = df_clean.drop(df_clean.index[0])
df_clean = df_clean.drop(np.nan, axis=1)

In [107]:
display_df(df_clean)

course_id  user_id registered viewed explored  \
1       HarvardX/PH525.1x/1T2018     7940      False           False   
2       HarvardX/PH525.1x/1T2018     7940      False           False   
3       HarvardX/PH525.1x/1T2018     7940      False           False   
4       HarvardX/PH525.1x/1T2018     7940      False           False   
5       HarvardX/PH525.1x/1T2018     7940      False           False   
...                          ...      ...        ...    ...      ...   
661482    HarvardX/CS50x/2014_T1  1934738       True  False            
661483    HarvardX/CS50x/2014_T1  1934738       True  False            
661484    HarvardX/CS50x/2014_T1  1934738       True  False            
661485    HarvardX/CS50x/2014_T1  1934738       True  False            
661486    HarvardX/CS50x/2014_T1  1934738       True  False            

       certified      completed             ip        cc_by_ip   countryLabel  \
1          False  81.108.107.58             GB  United Kingdom         Europe   
2          False  81.108.107.58             GB  United Kingdom         Europe   
3          False  81.108.107.58             GB  United Kingdom         Europe   
4          False  81.108.107.58             GB  United Kingdom         Europe   
5          False  81.108.107.58             GB  United Kingdom         Europe   
...          ...            ...            ...             ...            ...   
661482     False          False  71.41.229.179              US  United States   
661483     False          False  71.41.229.179              US  United States   
661484     False          False  71.41.229.179              US  United States   
661485     False          False  71.41.229.179              US  United States   
661486     False          False  71.41.229.179              US  United States   

            continent    city         region subdivision       postalCode  \
1       Middlesbrough     MDB  Middlesbrough              Northern Europe   
2       Middlesbrough     MDB  Middlesbrough              Northern Europe   
3       Middlesbrough     MDB  Middlesbrough              Northern Europe   
4       Middlesbrough     MDB  Middlesbrough              Northern Europe   
5       Middlesbrough     MDB  Middlesbrough              Northern Europe   
...               ...     ...            ...         ...              ...   
661482       Americas  Dallas             TX       Texas            75201   
661483       Americas  Dallas             TX       Texas            75201   
661484       Americas  Dallas             TX       Texas            75201   
661485       Americas  Dallas             TX       Texas            75201   
661486       Americas  Dallas             TX       Texas            75201   

          un_major_region  un_economic_group un_developing_nation  \
1       Developed regions                                           
2       Developed regions                                           
3       Developed regions                                           
4       Developed regions                                           
5       Developed regions                                           
...                   ...                ...                  ...   
661482   Northern America  Developed regions                        
661483   Northern America  Developed regions                        
661484   Northern America  Developed regions                        
661485   Northern America  Developed regions                        
661486   Northern America  Developed regions                        

       un_special_region latitude longitude LoE   YoB gender grade  \
1                54.5728  -1.1628                              0.7   
2                54.5728  -1.1628                              0.7   
3                54.5728  -1.1628                              0.7   
4                54.5728  -1.1628                              0.7   
5                54.5728  -1.1628                              0.7   
...        

In [93]:
stats_unique(df_clean, df_clean.columns)

Unique Count
course_id                      21
user_id                     49144
registered                      3
viewed                          3
explored                        3
certified                       2
completed                   44101
ip                            799
cc_by_ip                      295
countryLabel                   96
continent                    6509
city                         1145
region                       1646
subdivision                  9055
postalCode                    282
un_major_region                21
un_economic_group               4
un_developing_nation            4
un_special_region            9821
latitude                    10569
longitude                     443
LoE                           115
YoB                            59
gender                        105
grade                           8
passing_grade               47896
start_time                  33443
first_event                 40371
last_event                   3238
nevents                       267
ndays_act                     873
nplay_video                   111
nchapters                      68
nforum_posts                   43
nforum_votes                    6
nforum_endorsed                33
nforum_threads                 66
nforum_comments                 5
nforum_pinned                   3
roles                         107
nprogcheck                    578
nproblem_check                213
nforum_events                   5
mode                            6
is_active                   22286
cert_created_date           21901
cert_modified_date            608
cert_status                   607
NaN                             1

In [94]:
stats_NaN(df_clean)

NaN Ratio
course_id                       0
user_id                         0
registered                      0
viewed                0.000270603
explored              0.000498877
certified             0.000757386
completed              0.00102497
ip                     0.00126231
cc_by_ip               0.00155408
countryLabel           0.00181712
continent              0.00207563
city                   0.00235379
region                 0.00264405
subdivision            0.00291767
postalCode             0.00316258
un_major_region        0.00343167
un_economic_group      0.00372192
un_developing_nation    0.0039759
un_special_region      0.00421929
latitude               0.00446117
longitude              0.00468339
LoE                      0.004954
YoB                    0.00520948
gender                 0.00549067
grade                  0.00576883
passing_grade          0.00600617
start_time             0.00624805
first_event            0.00652017
last_event              0.0067424
nevents                0.00698427
ndays_act              0.00722464
nplay_video            0.00749978
nchapters               0.0077462
nforum_posts           0.00801226
nforum_votes           0.00826473
nforum_endorsed        0.00856707
nforum_threads         0.00882407
nforum_comments        0.00909619
nforum_pinned          0.00936528
roles                  0.00962076
nprogcheck             0.00989892
nproblem_check         0.00989892
nforum_events          0.00989892
mode                   0.00989892
is_active              0.00989892
cert_created_date      0.00989892
cert_modified_date     0.00989892
cert_status            0.00989892
NaN                      0.986152